src/main/java/org/apache/sysml/runtime/instructions/cp/ParamservBuiltinCPInstruction.java

```diff
@@ -95,44 +150,92 @@ public void processInstruction(ExecutionContext ec) {
 					String.format("Paramserv function: The argument '%s' could not be less than or equal to 0.",
 							PS_EPOCHS));
 		}
-		long batchSize = getBatchSize();
+
+		// Create an error handler
+		PSErrorHandler handler = new PSErrorHandler();
 
 		// Create the parameter server
 		ListObject model = ec.getListObject(getParam(PS_MODEL));
-		ListObject hyperParams = getHyperParams(ec);
-		ParamServer ps = createPS(mode, aggFunc, freq, updateType, workerNum, model, ec, hyperParams);
+		_ps = createPS(mode, aggFunc, freq, updateType, workerNum, model, aggServiceEC, handler);
 
 		// Create the local workers
-		List<LocalPSWorker> workers = IntStream.range(0, workerNum)
-				.mapToObj(i -> new LocalPSWorker((long) i, updFunc, freq, epochs, batchSize, hyperParams, ec, ps))
+		List<LocalPSWorker> workers = IntStream.range(0, workerNum).mapToObj(
+				i -> new LocalPSWorker((long) i, updFunc, freq, epochs, getBatchSize(), workerECs.get(i), _ps))
 				.collect(Collectors.toList());
 
 		// Do data partition
 		doDataPartition(ec, workers);
 
 		// Create the worker threads
-		List<Thread> threads = workers.stream().map(Thread::new).collect(Collectors.toList());
+		workers.forEach(w -> CompletableFuture.runAsync(w, _es).exceptionally(handler));
+
+		// Wait for the worker finishing
+		_es.shutdown();
```

Comment:
Ok, it will be more simple to handle the error in try/catch. And in fact, I'd like to leverage the feature of CompletableFuture which can combine all the futures instead of joining the thread one by one so that we could avoid to have some blocking case, i.e., if the second worker failed, maybe the main thread will always be blocked in joining the first worker but the first worker as well as the agg service are also blocked due to some reason. Thus, it will fall into a case never stopped.
Alternatively, I'd like to use:
```
		try {
			CompletableFuture[] futures = es.invokeAll(workers).stream().map(ParamservUtils::makeCompletableFuture)
					.toArray(CompletableFuture[]::new);
			CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futures);
			combinedFuture.get();
		} catch (InterruptedException | ExecutionException e) {
			throw new DMLRuntimeException("ParamservBuiltinCPInstruction: some error occurred: ", e);
		} finally {
			// shutdown the workers thread pool
			es.shutdownNow();
		}

```